string_manipulation.rb

```diff
@@ -1,30 +1,97 @@
 # A method to reverse a string in place.
 def string_reverse(my_string)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  j = my_string.length - 1
+  while i < j do
+    temp = my_string[i]
+    my_string[i] = my_string[j]
+    my_string[j] = temp
+    i += 1
+    j -=1
+  end
+  return my_string
 end
 
 # A method to reverse each word in a sentence, in place.
 def reverse_words(my_words)
-  puts "NOT IMPLEMENTED"
+  index = 0
+  i = 0
+  while index < my_words.length
+    until my_words[index] == " " || (index == my_words.length)
+      index += 1
+    end
+    j = index - 1
+    while i < j do
+      temp = my_words[i]
+      my_words[i] = my_words[j]
+      my_words[j] = temp
+      i += 1
+      j -= 1
+    end
+    index = index + 1
+    i = index
+  end
+  return my_words
 end
 
 # A method to reverse the words in a sentence, in place.
 def reverse_sentence(my_sentence)
-  puts "NOT IMPLEMENTED"
+  string_reverse(my_sentence)
+  reverse_words(my_sentence)
+  return my_sentence
 end
 
 # A method to check if the input string is a palindrome.
 # Return true if the string is a palindrome. Return false otherwise.
 def palindrome_check(my_phrase)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  j = my_phrase.length - 1
+  while i < j
+    if my_phrase[i] == " "
+      i += 1
+    elsif my_phrase[j] == " "
+      j -= 1
+    end
+    if my_phrase[i] != my_phrase[j]
+      return false
+    end
+    i += 1
+    j -= 1
+  end
   return true
 end
 
 # A method that updates the string by replacing consecutive repeating characters
 # with a number representing the frequncy. The replacement is done only if the
 # string length will get reduced by the process.
 def encode_repeating(my_string)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  counter = 0
+  index = 0
+  while i < my_string.length
+    while my_string[i] == my_string[i+1]
+      counter += 1
+      i += 1
+    end
+    # three or more consecutive repeating characters
+    if counter > 1
+      my_string[index..index+1] = "#{my_string[i]}#{counter+1}"
+      index += 2
+    # two repeating characters
+    elsif counter == 1
+      my_string[index..index+1] = my_string[i-1..i]

```

Comment:
Using ranges with Ruby strings creates temporary sub-strings, thereby using more memory. Could you do this explicitly by copying my_string[i-1] to be the value of my_string[index] and then copy the value at i to be the value at index+1? That won't create temporary substrings.